// @azure/static-web-apps-cli/dist/cli/index.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const commander_1 = __importDefault(require("commander"));
const update_notifier_1 = require("update-notifier");
const index_1 = require("./index");
const pkg = require("../../package.json");
const originalConsoleError = console.error;
const removeColors = (str) => str.replace(/[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g, "");
describe("cli", () => {
    beforeEach(() => {
        // Throws CommanderError instead of exiting after showing version or error
        commander_1.default.exitOverride();
        console.error = jest.fn();
    });
    afterAll(() => {
        console.error = originalConsoleError;
    });
    it("should print version", async () => {
        expect(index_1.run(["node", "swa", "-v"])).rejects.toThrow(pkg.version);
    });
    it("should notify of newer version", async () => {
        // Forces an update notification, bypassing regular checks
        process.stdout.isTTY = true;
        spyOn(update_notifier_1.UpdateNotifier.prototype, "check").and.callFake(function () {
            this.shouldNotifyInNpmScript = true;
            this.update = {
                current: "0.1.0",
                latest: "1.0.0",
                name: pkg.name,
            };
        });
        const orginalNotify = update_notifier_1.UpdateNotifier.prototype.notify;
        spyOn(update_notifier_1.UpdateNotifier.prototype, "notify").and.callFake(function (options = {}) {
            options.defer = false;
            orginalNotify.bind(this)(options);
        });
        await expect(index_1.run(["node", "swa", "-v"])).rejects.toThrow(pkg.version);
        expect(removeColors(console.error.mock.calls[0].join("\n"))).toMatchInlineSnapshot(`
      "
         ╭────────────────────────────────────────────────────╮
         │                                                    │
         │          Update available 0.1.0 → 1.0.0            │
         │   Run npm i @azure/static-web-apps-cli to update   │
         │                                                    │
         ╰────────────────────────────────────────────────────╯
      "
    `);
    });
});
