// @azure/static-web-apps-cli/dist/core/utils/net.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.hostnameToIpAdress = exports.parsePort = exports.response = exports.address = exports.parseUrl = exports.validateDevServerConfig = exports.isHttpUrl = exports.isAcceptingTcpConnections = void 0;
const chalk_1 = __importDefault(require("chalk"));
const net_1 = __importDefault(require("net"));
const ora_1 = __importDefault(require("ora"));
const wait_on_1 = __importDefault(require("wait-on"));
const logger_1 = require("./logger");
/**
 * Check if a given remote address and port are accepting TCP connections.
 * @param Object host and port of the server to check.
 * @returns True if the given server is accepting TCP connections. False otherwise.
 */
function isAcceptingTcpConnections({ host = "127.0.0.1", port }) {
    return new Promise((resolve) => {
        const socket = net_1.default.createConnection(port, host);
        socket
            .once("error", () => {
            resolve(false);
            socket.end();
        })
            .once("connect", () => {
            resolve(true);
            socket.end();
        });
    });
}
exports.isAcceptingTcpConnections = isAcceptingTcpConnections;
/**
 * Check if a given URL string is a valid URL.
 * @param url The URL string to check.
 * @returns True if the URL string is a valid URL. False otherwise.
 */
function isHttpUrl(url) {
    try {
        const uri = new URL(url);
        return uri.protocol.startsWith("http") || uri.protocol.startsWith("ws");
    }
    catch {
        return false;
    }
}
exports.isHttpUrl = isHttpUrl;
/**
 * Checks if a given server is up and accepting connection.
 * @param context An HTTP URL.
 * @param timeout Maximum time in ms to wait before exiting with failure (1) code,
  default Infinity.
 */
async function validateDevServerConfig(context, timeout) {
    let { hostname, port } = parseUrl(context);
    try {
        const appListening = await isAcceptingTcpConnections({ port, host: hostname });
        if (appListening === false) {
            const spinner = ora_1.default();
            try {
                spinner.start(`Waiting for ${chalk_1.default.green(context)} to be ready`);
                await wait_on_1.default({
                    resources: [`tcp:${hostname}:${port}`],
                    delay: 1000,
                    interval: 100,
                    simultaneous: 1,
                    timeout,
                    tcpTimeout: 1000,
                    window: 1000,
                    strictSSL: false,
                    verbose: false, // force disable verbose logs even if SWA_CLI_DEBUG is enabled
                });
                spinner.succeed(`Connected to ${chalk_1.default.green(context)} successfully`);
                spinner.clear();
            }
            catch (err) {
                spinner.fail();
                logger_1.logger.error(`Could not connect to "${context}". Is the server up and running?`);
                process.exit(-1);
            }
        }
    }
    catch (err) {
        if (err.message.includes("EACCES")) {
            logger_1.logger.error(`Port "${port}" cannot be used. You might need elevated or admin privileges. Or, use a valid port: 1024 to 49151`);
        }
        else {
            logger_1.logger.error(err.message);
        }
        process.exit(-1);
    }
}
exports.validateDevServerConfig = validateDevServerConfig;
/**
 * Parse a given URL and return its protocol, port, host and hostname.
 * @param url The URL string to check.
 * @returns Protocol, port, host and hostname extracted from the URL.
 */
function parseUrl(url) {
    const { protocol, port, host, hostname } = new URL(url);
    return {
        protocol,
        port: Number(port),
        host,
        hostname,
    };
}
exports.parseUrl = parseUrl;
/**
 * Construct a valid URL string from a host, port and protocol.
 * @param host A host address.
 * @param port (optional) A host port.
 * @param protocol (optional) A host protocol.
 * @throws {Error} if the URL is malformed.
 * @returns
 */
function address(host, port = 80, protocol = `http`) {
    if (!host) {
        throw new Error(`Host value is not set`);
    }
    let url = port === 80 ? `${protocol}://${host}` : `${protocol}://${host}:${port}`;
    if (isHttpUrl(url)) {
        return url;
    }
    else {
        throw new Error(`Address: ${url} is malformed!`);
    }
}
exports.address = address;
function response({ status, headers, cookies, body = "" }) {
    if (typeof status !== "number") {
        throw Error("TypeError: status code must be a number.");
    }
    body = body || null;
    const res = {
        status,
        cookies,
        headers: {
            status,
            "Content-Type": "application/json",
            ...headers,
        },
        body,
    };
    return res;
}
exports.response = response;
function parsePort(port) {
    const portNumber = parseInt(port, 10);
    if (isNaN(portNumber)) {
        logger_1.logger.error(`Port "${port}" is not a number.`, true);
    }
    else {
        if (portNumber < 1024 || portNumber > 49151) {
            logger_1.logger.error(`Port "${port}" is out of range. Allowed ports are from 1024 to 49151.`, true);
        }
    }
    return portNumber;
}
exports.parsePort = parsePort;
function hostnameToIpAdress(hostnameOrIpAddress) {
    if (hostnameOrIpAddress === "localhost") {
        return "127.0.0.1";
    }
    return hostnameOrIpAddress;
}
exports.hostnameToIpAdress = hostnameToIpAdress;
